packages/trackerless-network/src/logic/RandomGraphNode.ts

```diff
@@ -0,0 +1,297 @@
+import { EventEmitter } from 'eventemitter3'
+import {
+    DhtNode,
+    PeerDescriptor,
+    DhtPeer,
+    ListeningRpcCommunicator,
+    ITransport,
+    ConnectionLocker,
+    keyFromPeerDescriptor
+} from '@streamr/dht'
+import {
+    StreamMessage,
+    LeaveStreamNotice,
+    MessageRef,
+    StreamMessageType,
+    GroupKeyRequest
+} from '../proto/packages/trackerless-network/protos/NetworkRpc'
+import { PeerList } from './PeerList'
+import { NetworkRpcClient } from '../proto/packages/trackerless-network/protos/NetworkRpc.client'
+import { RemoteRandomGraphNode } from './RemoteRandomGraphNode'
+import { INetworkRpc } from '../proto/packages/trackerless-network/protos/NetworkRpc.server'
+import { DuplicateMessageDetector } from './DuplicateMessageDetector'
+import { Logger } from '@streamr/utils'
+import { toProtoRpcClient } from '@streamr/proto-rpc'
+import { IHandshaker } from './neighbor-discovery/Handshaker'
+import { Propagation } from './propagation/Propagation'
+import { INeighborFinder } from './neighbor-discovery/NeighborFinder'
+import { INeighborUpdateManager } from './neighbor-discovery/NeighborUpdateManager'
+import { PeerIDKey } from '@streamr/dht/dist/src/helpers/PeerID'
+import { StreamNodeServer } from './StreamNodeServer'
+import { IStreamNode } from './IStreamNode'
+import { ProxyStreamConnectionServer } from './proxy/ProxyStreamConnectionServer'
+import { markAndCheckDuplicate } from './utils'
+
+export interface Events {
+    message: (message: StreamMessage) => void
+    targetNeighborConnected: (stringId: string) => void
+    nearbyContactPoolIdAdded: () => void
+}
+
+export interface StrictRandomGraphNodeConfig {
+    randomGraphId: string
+    layer1: DhtNode
+    P2PTransport: ITransport
+    connectionLocker: ConnectionLocker
+    ownPeerDescriptor: PeerDescriptor
+    peerViewSize: number
+    nearbyContactPool: PeerList
+    randomContactPool: PeerList
+    targetNeighbors: PeerList
+    handshaker: IHandshaker
+    neighborFinder: INeighborFinder
+    neighborUpdateManager: INeighborUpdateManager
+    propagation: Propagation
+    rpcCommunicator: ListeningRpcCommunicator
+    numOfTargetNeighbors: number
+    maxNumberOfContacts: number
+    minPropagationTargets: number
```

Comment:
Is this the minimum number of neighbors which we need to propagate each message? When we'll have proxy connections and temporary connections, does this include/exclude those?

Reply:
yes, and yes it includes proxy connections and temporary connections

Reply:
So if `minPropapagationTargets` is `2` and a node has one proxy connection and one temporary connection, it doesn't have to propagate to the "normal" connections at all?

Reply:
No, if there is only one connection in total, messages will be kept in buffer until a second connections is formed